File: leetcode/13.cpp
#include <unordered_map>
#include <string>

using namespace std;

class Solution {
public:
  int romanToInt(string s) {
    unordered_map<char, int> cache({{'I', 1},
                                    {'V', 5},
                                    {'X', 10},
                                    {'L', 50},
                                    {'C', 100},
                                    {'D', 500},
                                    {'M', 1000}});
    int val = 0;
    int prev = 0;
    for (int i = s.size() - 1; i >= 0; --i) {
      int cur = cache[s[i]];
      val += (cur >= prev ? cur : -cur);
      prev = cur;
    }
    return val;
  }
};

File: leetcode/13.py
class Solution:
    def romanToInt(self, s: str) -> int:
        signs = {"M": 1000, "CM": 900, "D": 500, "CD": 400, "C": 100, "XC": 90, "L": 50,
                 "XL": 40, "X": 10, "IX": 9, "V": 5, "IV": 4, "I": 1}
        num = 0
        n = len(s)
        pos = 0

        while pos < n:
            if s[pos:pos + 2] in signs:
                num += signs[s[pos:pos + 2]]
                pos += 2
            else:
                num += signs[s[pos:pos + 1]]
                pos += 1

        return num
